core/embed/rust/src/ui/layout_bolt/component/keyboard/pin.rs

```diff
@@ -362,81 +345,105 @@ impl PinDots {
         &self.digits
     }
 
-    fn render_digits<'s>(&self, area: Rect, target: &mut impl Renderer<'s>) {
+    fn render_shown<'s>(&self, area: Rect, target: &mut impl Renderer<'s>) {
+        // Make sure the pin should be shown
+        debug_assert_eq!(self.display_style, DisplayStyle::Shown);
+
         let center = area.center() + Offset::y(fonts::FONT_MONO.text_height() / 2);
         let right =
-            center + Offset::x(fonts::FONT_MONO.text_width("0") * (MAX_VISIBLE_DOTS as i16) / 2);
-        let digits = self.digits.len();
+            center + Offset::x(fonts::FONT_MONO.text_width("0") * (MAX_SHOWN_LEN as i16) / 2);
+        let pin_len = self.digits.len();
 
-        if digits <= MAX_VISIBLE_DOTS {
-            shape::Text::new(center, &self.digits, fonts::FONT_MONO)
+        if pin_len <= MAX_SHOWN_LEN {
+            Text::new(center, &self.digits, fonts::FONT_MONO)
                 .with_align(Alignment::Center)
                 .with_fg(self.style.text_color)
                 .render(target);
         } else {
-            let offset: usize = digits.saturating_sub(MAX_VISIBLE_DIGITS);
-            shape::Text::new(right, &self.digits[offset..], fonts::FONT_MONO)
+            let offset = pin_len.saturating_sub(MAX_SHOWN_LEN);
+            Text::new(right, &self.digits[offset..], fonts::FONT_MONO)
                 .with_align(Alignment::End)
                 .with_fg(self.style.text_color)
                 .render(target);
         }
     }
 
-    fn render_dots<'s>(&self, last_digit: bool, area: Rect, target: &mut impl Renderer<'s>) {
+    fn render_hidden<'s>(&self, area: Rect, target: &mut impl Renderer<'s>) {
+        debug_assert_ne!(self.display_style, DisplayStyle::Shown);
+
         let mut cursor = self.size().snap(area.center(), Alignment2D::CENTER);
 
-        let digits = self.digits.len();
-        let dots_visible = digits.min(MAX_VISIBLE_DOTS);
-        let step = Self::DOT + Self::PADDING;
+        let pin_len = self.digits.len();
+        let last_digit = self.display_style == DisplayStyle::LastOnly;
+        let step = Self::ICON_WIDTH + Self::ICON_SPACING;
+
+        // Render only when there are characters
+        if pin_len > 0 {
```

Comment:
nit: consider negating the check here
```suggestion
        if pin_len == 0 {
            return;
        }
```

Reply:
https://github.com/trezor/trezor-firmware/pull/5317/commits/a23a39e3510b1e78339af54b3386cd01fa6701c1